disallow RNGs in size declarations

We should not allow RNGs in size declarations.  As is, the following program fails during C++ compilation:

```
generated quantities {
  int a[bernoulli_rng(0.5)];
}
```

Replacing the `bernoulli_rng(0.5)` with `1` works just fine.

In RStan, here's the error trying to compile the model:

```
> stan_model('random.stan')
Error in compileCode(f, code, language = language, verbose = verbose) : 
  Compilation ERROR, function(s)/method(s) not created! In file included from <built-in>:1:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/StanHeaders/include/stan/math/prim/mat/fun/Eigen.hpp:4:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/Eigen/Dense:1:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/Eigen/Core:531:
/Library/Frameworks/R.framework/Versions/3.5/Resources/library/RcppEigen/include/Eigen/src/Core/util/ReenableStupidWarnings.h:10:30: warning: pragma diagnostic pop could not pop, no matching push [-Wunknown-pragmas]
    #pragma clang diagnostic pop
                             ^
In file included from <built-in>:1:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/StanHeaders/include/stan/math/prim/mat/fun/Eigen.hpp:4:
In file included from /Library/Frameworks/R.framework/Versions/3.
In addition: Warning message:
In system(cmd, intern = !verbose) :
  running command '/Library/Frameworks/R.framework/Resources/bin/R CMD SHLIB file14e938ca225d.cpp 2> file14e938ca225d.cpp.err.txt' had status 1
Error in sink(type = "output") : invalid connection
```

--- a/src/frontend/Semantic_check.ml
+++ b/src/frontend/Semantic_check.ml
@@ -51,6 +51,7 @@ let vm = Symbol_table.initialize ()
    used for error reporting. *)
 type context_flags_record =
   { current_block: originblock
+  ; in_toplevel_decl: bool
   ; in_fun_def: bool
   ; in_returning_fun_def: bool
   ; in_rng_fun_def: bool
@@ -254,11 +255,13 @@ let semantic_check_fn_target_plus_equals cf ~loc id =
     else ok ())
 
 (** Rng functions cannot be used in Tp or Model and only
-    in funciton defs with the right suffix
+    in function defs with the right suffix
 *)
 let semantic_check_fn_rng cf ~loc id =
   Validate.(
-    if
+    if String.is_suffix id.name ~suffix:"_rng" && cf.in_toplevel_decl then
+      Semantic_error.invalid_decl_rng_fn loc |> error
+    else if
       String.is_suffix id.name ~suffix:"_rng"
       && ( (cf.in_fun_def && not cf.in_rng_fun_def)
          || cf.current_block = TParam || cf.current_block = Model )
@@ -463,6 +466,8 @@ let semantic_check_variable cf loc id =
             (calculate_autodifftype cf MathLibrary UMathLibraryFunction)
           ~type_:UMathLibraryFunction ~loc
         |> ok
+    | Some ((Param | TParam | GQuant), _) when cf.in_toplevel_decl ->
+        Semantic_error.non_data_variable_size_decl loc |> error
     | Some (originblock, type_) ->
         mk_typed_expression ~expr:(Variable id)
           ~ad_level:(calculate_autodifftype cf originblock type_)
@@ -1408,24 +1413,6 @@ and semantic_check_block ~loc ~cf stmts =
         mk_typed_statement ~stmt:(Block xs) ~return_type ~loc ))
 
 (* -- Variable Declarations ------------------------------------------------- *)
-and semantic_check_size_decl ~loc is_global sized_ty =
-  let not_ptq e =
-    match e.emeta.ad_level with AutoDiffable -> false | _ -> true
-  in
-  let rec check_sizes_data_only = function
-    | SizedType.SVector e -> not_ptq e
-    | SRowVector e -> not_ptq e
-    | SMatrix (e1, e2) -> not_ptq e1 && not_ptq e2
-    | SArray (sized_ty, e) when not_ptq e -> check_sizes_data_only sized_ty
-    | SArray _ -> false
-    | _ -> true
-  in
-  (* Sizes must be of level at most data. *)
-  Validate.(
-    if is_global && not (check_sizes_data_only sized_ty) then
-      Semantic_error.non_data_variable_size_decl loc |> error
-    else ok ())
-
 and semantic_check_var_decl_bounds ~loc is_global sized_ty trans =
   let is_real {emeta; _} = emeta.type_ = UReal in
   let is_valid_transformation =
@@ -1471,10 +1458,7 @@ and semantic_check_var_decl_initial_value ~loc ~cf id init_val_opt =
 
 and semantic_check_var_decl ~loc ~cf sized_ty trans id init is_global =
   let checked_stmt =
-    Validate.(
-      semantic_check_sizedtype cf sized_ty
-      >>= fun ust ->
-      semantic_check_size_decl ~loc is_global ust |> map ~f:(fun _ -> ust))
+    semantic_check_sizedtype {cf with in_toplevel_decl= is_global} sized_ty
   in
   let checked_trans = semantic_check_transformation cf trans in
   Validate.(
@@ -1482,8 +1466,6 @@ and semantic_check_var_decl ~loc ~cf sized_ty trans id init is_global =
     |> apply_const (semantic_check_identifier id)
     |> apply_const (check_fresh_variable id false)
     >>= fun (ust, utrans) ->
-    semantic_check_size_decl ~loc is_global ust
-    >>= fun _ ->
     let ut = unsizedtype_of_sizedtype ust in
     Symbol_table.enter vm id.name (cf.current_block, ut) ;
     semantic_check_var_decl_initial_value ~loc ~cf id init
@@ -1765,6 +1747,7 @@ let semantic_check_program
   unsafe_clear_symbol_table vm ;
   let cf =
     { current_block= Functions
+    ; in_toplevel_decl= false
     ; in_fun_def= false
     ; in_returning_fun_def= false
     ; in_rng_fun_def= false

--- a/src/frontend/Semantic_error.ml
+++ b/src/frontend/Semantic_error.ml
@@ -241,6 +241,7 @@ end
 module ExpressionError = struct
   type t =
     | InvalidMapRectFn of string
+    | InvalidSizeDeclRng
     | InvalidRngFunction
     | ConditionalNotationNotAllowed
     | ConditioningRequired
@@ -253,6 +254,10 @@ module ExpressionError = struct
           "Mapped function cannot be an _rng or _lp function, found function \
            name: %s"
           fn_name
+    | InvalidSizeDeclRng ->
+        Fmt.pf ppf
+          "Random number generators are not allowed in top level size \
+           declarations."
     | InvalidRngFunction ->
         Fmt.pf ppf
           "Random number generators are only allowed in transformed data \
@@ -506,6 +511,9 @@ let ident_not_in_scope loc name =
 
 let invalid_map_rect_fn loc name =
   ExpressionError (loc, ExpressionError.InvalidMapRectFn name)
+
+let invalid_decl_rng_fn loc =
+  ExpressionError (loc, ExpressionError.InvalidSizeDeclRng)
 
 let invalid_rng_fn loc =
   ExpressionError (loc, ExpressionError.InvalidRngFunction)

--- a/src/frontend/Semantic_error.mli
+++ b/src/frontend/Semantic_error.mli
@@ -73,6 +73,7 @@ val ident_is_stanmath_name : Location_span.t -> string -> t
 val ident_in_use : Location_span.t -> string -> t
 val ident_not_in_scope : Location_span.t -> string -> t
 val invalid_map_rect_fn : Location_span.t -> string -> t
+val invalid_decl_rng_fn : Location_span.t -> t
 val invalid_rng_fn : Location_span.t -> t
 val conditional_notation_not_allowed : Location_span.t -> t
 val conditioning_required : Location_span.t -> t

--- a/test/integration/bad/data_index/rng_index1.stan
+++ b/test/integration/bad/data_index/rng_index1.stan
@@ -1,0 +1,7 @@
+parameters {
+  real y[3];
+
+}
+transformed parameters {
+  real z[poisson_rng(10)];
+}

--- a/test/integration/bad/data_index/stanc.expected
+++ b/test/integration/bad/data_index/stanc.expected
@@ -1,11 +1,11 @@
   $ ../../../../../install/default/bin/stanc non_data_index1.stan
 
-Semantic error in 'non_data_index1.stan', line 6, column 2 to column 18:
+Semantic error in 'non_data_index1.stan', line 6, column 14 to column 15:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    real z[size(y)];
-           ^
+                       ^
      7:  }
    -------------------------------------------------
 
@@ -13,12 +13,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index10.stan
 
-Semantic error in 'non_data_index10.stan', line 6, column 2 to column 33:
+Semantic error in 'non_data_index10.stan', line 6, column 27 to column 28:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    cholesky_factor_cov[size(y)] z;
-           ^
+                                    ^
      7:  }
    -------------------------------------------------
 
@@ -26,12 +26,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index11.stan
 
-Semantic error in 'non_data_index11.stan', line 6, column 2 to column 36:
+Semantic error in 'non_data_index11.stan', line 6, column 30 to column 31:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    cholesky_factor_cov[3, size(y)] z;
-           ^
+                                       ^
      7:  }
    -------------------------------------------------
 
@@ -39,12 +39,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index12.stan
 
-Semantic error in 'non_data_index12.stan', line 6, column 2 to column 34:
+Semantic error in 'non_data_index12.stan', line 6, column 28 to column 29:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    cholesky_factor_corr[size(y)] z;
-           ^
+                                     ^
      7:  }
    -------------------------------------------------
 
@@ -52,12 +52,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index13.stan
 
-Semantic error in 'non_data_index13.stan', line 6, column 2 to column 25:
+Semantic error in 'non_data_index13.stan', line 6, column 19 to column 20:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    corr_matrix[size(y)] z;
-           ^
+                            ^
      7:  }
    -------------------------------------------------
 
@@ -65,12 +65,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index14.stan
 
-Semantic error in 'non_data_index14.stan', line 6, column 2 to column 24:
+Semantic error in 'non_data_index14.stan', line 6, column 18 to column 19:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    cov_matrix[size(y)] z;
-           ^
+                           ^
      7:  }
    -------------------------------------------------
 
@@ -78,12 +78,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index2.stan
 
-Semantic error in 'non_data_index2.stan', line 6, column 2 to column 20:
+Semantic error in 'non_data_index2.stan', line 6, column 14 to column 15:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    vector[size(y)] z;
-           ^
+                       ^
      7:  }
    -------------------------------------------------
 
@@ -91,12 +91,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index3.stan
 
-Semantic error in 'non_data_index3.stan', line 6, column 2 to column 24:
+Semantic error in 'non_data_index3.stan', line 6, column 18 to column 19:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    row_vector[size(y)] z;
-           ^
+                           ^
      7:  }
    -------------------------------------------------
 
@@ -104,12 +104,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index4.stan
 
-Semantic error in 'non_data_index4.stan', line 5, column 2 to column 23:
+Semantic error in 'non_data_index4.stan', line 5, column 14 to column 15:
    -------------------------------------------------
      3:  }
      4:  transformed parameters {
      5:    matrix[size(y), 2] m;
-           ^
+                       ^
      6:  }
    -------------------------------------------------
 
@@ -117,12 +117,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index5.stan
 
-Semantic error in 'non_data_index5.stan', line 6, column 2 to column 23:
+Semantic error in 'non_data_index5.stan', line 6, column 17 to column 18:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    matrix[2, size(y)] m;
-           ^
+                          ^
      7:  
      8:  }
    -------------------------------------------------
@@ -131,12 +131,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index6.stan
 
-Semantic error in 'non_data_index6.stan', line 6, column 2 to column 21:
+Semantic error in 'non_data_index6.stan', line 6, column 15 to column 16:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    simplex[size(y)] z;
-           ^
+                        ^
      7:  }
    -------------------------------------------------
 
@@ -144,12 +144,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index7.stan
 
-Semantic error in 'non_data_index7.stan', line 6, column 2 to column 21:
+Semantic error in 'non_data_index7.stan', line 6, column 15 to column 16:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    ordered[size(y)] z;
-           ^
+                        ^
      7:  }
    -------------------------------------------------
 
@@ -157,12 +157,12 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index8.stan
 
-Semantic error in 'non_data_index8.stan', line 6, column 2 to column 30:
+Semantic error in 'non_data_index8.stan', line 6, column 24 to column 25:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    positive_ordered[size(y)] z;
-           ^
+                                 ^
      7:  }
    -------------------------------------------------
 
@@ -170,14 +170,27 @@ Non-data variables are not allowed in top level size declarations.
 
   $ ../../../../../install/default/bin/stanc non_data_index9.stan
 
-Semantic error in 'non_data_index9.stan', line 6, column 2 to column 25:
+Semantic error in 'non_data_index9.stan', line 6, column 19 to column 20:
    -------------------------------------------------
      4:  }
      5:  transformed parameters {
      6:    unit_vector[size(y)] z;
-           ^
+                            ^
      7:  }
    -------------------------------------------------
 
 Non-data variables are not allowed in top level size declarations.
+
+  $ ../../../../../install/default/bin/stanc rng_index1.stan
+
+Semantic error in 'rng_index1.stan', line 6, column 9 to column 24:
+   -------------------------------------------------
+     4:  }
+     5:  transformed parameters {
+     6:    real z[poisson_rng(10)];
+                  ^
+     7:  }
+   -------------------------------------------------
+
+Random number generators are not allowed in top level size declarations.
 

--- a/test/integration/bad/stanc.expected
+++ b/test/integration/bad/stanc.expected
@@ -3207,15 +3207,29 @@ Matrix sizes must be of type int. Instead found type real.
 
   $ ../../../../install/default/bin/stanc var-decl-bad-1.stan
 
-Semantic error in 'var-decl-bad-1.stan', line 3, column 2 to column 18:
+Semantic error in 'var-decl-bad-1.stan', line 3, column 14 to column 15:
    -------------------------------------------------
      1:  parameters {
      2:    real x[3];
      3:    real y[size(x)];
-           ^
+                       ^
      4:  }
      5:  model {
    -------------------------------------------------
 
 Non-data variables are not allowed in top level size declarations.
+
+  $ ../../../../install/default/bin/stanc var-decl-bad-2.stan
+
+Semantic error in 'var-decl-bad-2.stan', line 3, column 9 to column 23:
+   -------------------------------------------------
+     1:  parameters {
+     2:    real x[3];
+     3:    real y[poisson_rng(3)];
+                  ^
+     4:  }
+     5:  model {
+   -------------------------------------------------
+
+Random number generators are not allowed in top level size declarations.
 

--- a/test/integration/bad/var-decl-bad-2.stan
+++ b/test/integration/bad/var-decl-bad-2.stan
@@ -1,0 +1,7 @@
+parameters {
+  real x[3];
+  real y[poisson_rng(3)];
+}
+model {
+  y ~ normal(0,1);
+}

